ddht/v5_1/alexandria/client.py

```diff
@@ -110,6 +131,93 @@ async def _request(
             )
         return response  # type: ignore
 
+    def subscribe(
+        self,
+        message_type: Type[TAlexandriaMessage],
+        endpoint: Optional[Endpoint] = None,
+        node_id: Optional[NodeID] = None,
+    ) -> AsyncContextManager[
+        trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]
+    ]:
+        return self.subscription_manager.subscribe(message_type, endpoint, node_id,)  # type: ignore
+
+    @asynccontextmanager
+    async def subscribe_request(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        request: AlexandriaMessage[Any],
+        response_message_type: Type[TAlexandriaMessage],
+        request_id: Optional[bytes],
+    ) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]]:
+        if request_id is None:
+            request_id = self.network.client.request_tracker.get_free_request_id(
+                node_id
+            )
+
+        self.logger.debug(
+            "Sending request: %s with request id %s", request, request_id.hex(),
+        )
+
+        send_channel, receive_channel = trio.open_memory_channel[
+            InboundMessage[TAlexandriaMessage]
+        ](256)
+
+        with self.request_tracker.reserve_request_id(node_id, request_id):
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(
```

Comment:
you should be able to do: `nursery.start_soon(self._manage_request_response, node_id, endpoint, ...)`

Reply:
This was a workaround (which I'll document) for an inadequacy in how `trio` types the `start_soon` function.

```
ddht/v5_1/alexandria/client.py:178: error: Argument 1 to "start_soon" of "Nursery" has incompatible type "Callable[[NodeID, Endpoint, AlexandriaMessage[Any], Type[AlexandriaMessage[Any]], SendChannel[InboundMessage[AlexandriaMessage[Any]]], bytes], Coroutine[Any, Any, None]]"; expected "Union[Callable[[], Awaitable[Any]], Callable[[Any], Awaitable[Any]], Callable[[Any, Any], Awaitable[Any]], Callable[[Any, Any, Any], Awaitable[Any]], Callable[[Any, Any, Any, Any], Awaitable[Any]], Callable[[VarArg(Any)], Awaitable[Any]]]"
```

Anything beyond 4 args isn't covered in the type hints.